common/prices.go

```diff
@@ -0,0 +1,352 @@
+				return price, err
+			}
+
+		case map[string]interface{}:
+			priceJSON = v
+		}
+
+	}
+
+	return -1, errors.New("path didn't result in lookup")
+}
+
+func fetchCoinbasePrice() (float64, error) {
+	return fetchAPIPrice("https://api.coinbase.com/v2/exchange-rates?currency=ZEC", []string{"data", "rates", "USD"})
+}
+
+func fetchCoinCapPrice() (float64, error) {
+	return fetchAPIPrice("https://api.coincap.io/v2/rates/zcash", []string{"data", "rateUsd"})
+}
+
+func fetchBinancePrice() (float64, error) {
+	return fetchAPIPrice("https://api.binance.com/api/v3/avgPrice?symbol=ZECUSDC", []string{"price"})
+}
+
+func fetchHistoricalCoingeckoPrice(ts time.Time) (float64, error) {
+	dt := ts.Format("02-01-2006") // dd-mm-yyyy
+	url := fmt.Sprintf("https://api.coingecko.com/api/v3/coins/zcash/history?date=%s", dt)
+
+	return fetchAPIPrice(url, []string{"market_data", "current_price", "usd"})
+}
+
+// calcMedian calculates the median of a sorted slice of numbers
+func calcMedian(inp []float64) (median float64) {
+	// For even numbers we add the two middle numbers
+	// and divide by two using the mean function above
+	// For odd numbers we just use the middle number
+	n := len(inp)
+	if n%2 == 0 {
+		return (inp[n/2-1] + inp[n/2]) / 2
+	} else {
+		return inp[n/2]
+	}
+}
+
+// fetchPriceFromWebAPI will fetch prices from multiple places, discard outliers and return the
+// concensus price. This function doesn't need the mutex.
+func fetchPriceFromWebAPI() (float64, error) {
+	// We'll fetch prices from all our endpoints, and use the median price from that
+	priceProviders := []func() (float64, error){fetchBinancePrice, fetchCoinCapPrice, fetchCoinbasePrice}
+
+	// Get all prices
+	prices := make([]float64, 0)
+	for _, provider := range priceProviders {
+		price, err := provider()
+		if err == nil {
+			Log.WithFields(logrus.Fields{
+				"method":   "CurrentPrice",
+				"provider": runtime.FuncForPC(reflect.ValueOf(provider).Pointer()).Name(),
+				"price":    price,
+			}).Info("Service")
+			prices = append(prices, price)
+		} else {
+			Log.WithFields(logrus.Fields{
+				"method":   "CurrentPrice",
+				"provider": runtime.FuncForPC(reflect.ValueOf(provider).Pointer()).Name(),
+				"error":    err,
+			}).Error("Service")
+		}
+	}
+	if len(prices) == 0 {
+		return -1, errors.New("no price providers are available")
+	}
+	sort.Float64s(prices)
+
+	// Get the median price
+	median := calcMedian(prices)
+
+	// Discard all values that are more than 20% outside the median
+	validPrices := make([]float64, 0)
+	for _, price := range prices {
+		if (math.Abs(price-median) / median) > 0.2 {
+			Log.WithFields(logrus.Fields{
+				"method": "CurrentPrice",
+				"error":  fmt.Sprintf("Discarding price (%.2f) because too far away from median (%.2f", price, median),
+			}).Error("Service")
+		} else {
+			validPrices = append(validPrices, price)
+		}
+	}
+
+	// At least 2 (valid) providers are required; we don't want to depend on just one
```

Comment:
Requiring at least two sources means that either of them can determine the price (since we have to assume they know the price from the other source, then they can choose whatever median they want).

I think at least three sources should be required. In that case no single source can choose an outlier price: a malicious/corrupted source can only cause the median to be any price between the other two.